rosidl_generator_py/rosidl_generator_py/__init__.py

```diff
@@ -0,0 +1,146 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+import os
+
+from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_cmake import expand_template
+from rosidl_cmake import get_newest_modification_time
+from rosidl_cmake import read_generator_arguments
+from rosidl_parser import parse_message_file
+from rosidl_parser import parse_service_file
+
+
+def generate_py(generator_arguments_file):
+    args = read_generator_arguments(generator_arguments_file)
+
+    template_dir = args['template_dir']
+    mapping_msgs = {
+        os.path.join(template_dir, 'msg.py.template'): '_%s.py',
+    }
+
+    mapping_srvs = {
+        os.path.join(template_dir, 'srv.py.template'): '_%s.py',
+    }
+    for template_file in mapping_msgs.keys():
+        assert os.path.exists(template_file)
+    for template_file in mapping_srvs.keys():
+        assert os.path.exists(template_file)
+
+    functions = {
+        'get_header_filename_from_msg_name': convert_camel_case_to_lower_case_underscore,
+        'value_to_py': value_to_py,
+    }
+    latest_target_timestamp = get_newest_modification_time(args['target_dependencies'])
+
+    modules = defaultdict(list)
+    base_path = os.path.basename(os.path.dirname(args['output_dir']))
+    for ros_interface_file in args['ros_interface_files']:
+        extension = os.path.splitext(ros_interface_file)[1]
+        subfolder = os.path.basename(os.path.dirname(ros_interface_file))
+        import_path = convert_path_to_import(os.path.join(base_path, subfolder))
+        if extension == '.msg':
+            spec = parse_message_file(args['package_name'], ros_interface_file)
+            mapping = mapping_msgs
+            type_name = spec.base_type.type
+        elif extension == '.srv':
+            spec = parse_service_file(args['package_name'], ros_interface_file)
+            mapping = mapping_srvs
+            type_name = spec.srv_name
+        else:
+            continue
+
+        modules[subfolder].append(type_name)
+        for template_file, generated_filename in mapping.items():
+            data = {'import_path': import_path, 'spec': spec, 'subfolder': subfolder}
+            data.update(functions)
+            generated_file = os.path.join(
+                args['output_dir'], subfolder, generated_filename % type_name)
+            expand_template(
+                template_file, data, generated_file,
+                minimum_timestamp=latest_target_timestamp)
+
+    open(os.path.join(args['output_dir'], '__init__.py'), 'w').close()
+    for module in modules:
+        with open(os.path.join(args['output_dir'], module, '__init__.py'), 'w') as f:
+            for type_ in modules[module]:
+                f.write('from %s._%s import %s\n' % (import_path, type_, type_))
+
+    return 0
+
+
+def convert_path_to_import(path):
```

Comment:
Why is this necessary? The package knows the import pattern precisely (`<package_name>.msg`) and can just construct this.